### PR TITLE
Config imports

### PR DESCRIPTION
**Pull request by plaindocs**

### CODE REVIEW

File: user/imports.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  - 2.6
+script: bundle exec rake
+```
+{: data-file=".travis.yml"}
+
+## Import key
+
+You can import up to 5 YAML sources/snippets into one `.travis.yml`, and you can of course reuse those snippets in multiple repositories.
+
+The format of the `import` source is `<account>/<repository>/<path>@<ref>` in whic `<ref>` can be any valid Git reference, such as a commit sha, branch name, or tag name. If you do not specify `<ref>`, Travis CI will fetch the latest version of the master branch in the repository.
```
Comment (MariadeAnton) [inline user/imports.md]:
```suggestion
The format of the `import` source is `<account>/<repository>/<path>@<ref>` in whic `<ref>` can be any valid Git reference, such as a commit sha, branch name, or tag name. If you do not specify `<ref>`, Travis CI will fetch the latest version of the default branch in the repository.
```

File: user/imports.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+```yaml
+# local imports fetch the same git commit ref
+import:
+- ./one.yml
+- ./other.yml
+```
+{: data-file=".travis.yml"}
+
+By default, configurations imported from the same repository, fetch the commit you’re currently building. This is intended to help while you are  creating and testing the shared configurations.
```
Comment (MariadeAnton) [inline user/imports.md]:
```suggestion
By default, configurations imported from the same repository, fetch the commit you’re currently building. This is intended to help while you are  creating and testing the shared configurations.
You can override this value by specifying a valid Git reference.
```
would the full "valid Git reference, such as a commit sha, branch name, or tag name" help here too @plaindocs ?

Comment (MariadeAnton) [reply user/imports.md]: There's a link to Git References too https://git-scm.com/book/en/v2/Git-Internals-Git-References but that may confuse things more?

Comment (plaindocs) [reply user/imports.md]: Yeah, might be worth rewording.

File: user/imports.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
+---
+title: Importing shared build configuration
+
+layout: en
+---
+
+The main source of configuration for your build is the `.travis.yml` file stored in your repository. Now you can import up to five configuration snippets into your `.travis.yml`, so you can update multiple repositories making only one change.
```
Comment (svenfuchs) [inline user/imports.md]: "Now you can" seems like a time reference that would not age well? :) I'd strike "now".

File: user/imports.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+By default, if the same keys are present in multiple imported snippets, they are overwritten with the last configuration imported. But you can also specify [deep_merge](#merge-mode).
+
+## Merge mode
+
+The default merge `mode` is `merge` which performs a shallow merge.
+This means that root level sections (keys) defined in your `.travis.yml` will be overwritten by root level sections (keys) that are also present in the imported yaml file.
+Merge modes are specific to each source.
+
+You can also use `deep_merge` and `replace`:
```
Comment (svenfuchs) [inline user/imports.md]: i would not mention the merge mode `replace` here. for imports there's not a single use case ... the importing config would always entirely replace the imported one, making the whole import pointless :)

File: user/imports.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+#
+- source: travis-ci/shared-configs/other.yml@main
+  mode: deep_merge
+```
+{: data-file=".travis.yml"}
+
+
+## Importing configurations from the same repository
+
+When importing configurations stored in the same repository as your `travis.yml`, you can ommit the `<account>/<repository>` and use only the path:
```
Comment (svenfuchs) [inline user/imports.md]: Maybe add something like 

> add a dot and slash `./` at the beginning to indicate that this is a local path

?

File: user/imports.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+import:
+- ./one.yml
+- ./other.yml
+```
+{: data-file=".travis.yml"}
+
+By default, configurations imported from the same repository, fetch the commit you’re currently building. This is intended to help while you are  creating and testing the shared configurations.
+You can override this value by specifying a valid Git reference.
+
+## Builds trigged by the application
```
Comment (svenfuchs) [inline user/imports.md]: I somewhat tripped over "the application" and wondered "which application?"

Maybe change this to "Manually triggered builds" or "Build triggered via API or UI"?

**Comment by Pezi777** [comment]
@svenfuchs and @joshk Can this pull request be merged?

File: user/imports.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
+---
+title: Importing shared build configuration
+
+layout: en
+---
+
+The main source of configuration for your build is the `.travis.yml` file stored in your repository. You can import up to five configuration snippets into your `.travis.yml`, so you can update multiple repositories making only one change.
```
Comment (MariadeAnton) [inline user/imports.md]:
```suggestion
The main source of configuration for your build is the `.travis.yml` file stored in your repository. You can import up to 25 configuration snippets into your `.travis.yml`, so you can update multiple repositories making only one change.
```
cc/ @svenfuchs

File: _includes/sidebar.html [hunk 8]
```diff
@@ -40,6 +40,8 @@ <h3>Getting Started</h3>
         <li><a href="/user/cron-jobs/">Cron Jobs</a></li>
         <li><a href="/user/common-build-problems/">Common Build Problems</a></li>
         <li><a href="https://github.com/travis-ci/travis#readme">Command Line Client</a></li>
+        <li><a href="/user/imports/">Configuration imports</a></li>
```
Comment (Pezi777) [inline _includes/sidebar.html]:
```suggestion
        <li><a href="/user/imports/">Configuration Imports</a></li>
```

File: user/config-imports.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
+---
+title: Importing shared build configuration
```
Comment (Pezi777) [inline user/config-imports.md]:
```suggestion
title: Importing Shared Build Configuration
```

File: user/config-imports.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+
+layout: en
+---
+
+The main source of configuration for your build is the `.travis.yml` file
+stored in your repository. You can import shared configuration snippets into
+your `.travis.yml` or [API build request payload](https://docs.travis-ci.com/user/triggering-builds/),
+so you can update your build configuration in multiple repositories making only
```
Comment (Pezi777) [inline user/config-imports.md]:
```suggestion
to update your build configuration in multiple repositories making only
```

File: user/config-imports.md [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+one change.
+
+Imported configs can themselves include other configs, making this feature very
+composable (cyclic imports will be skipped). You can import up to 25 build
+configuration snippets in total.
+
+
+> BETA Configuration imports is currently in beta. Please leave feedback on the [Community forum](https://travis-ci.community/c/early-releases).
```
Comment (Pezi777) [inline user/config-imports.md]:
```suggestion
> BETA The Configuration Imports feature is currently in beta. You can leave your feedback on the [Community forum](https://travis-ci.community/c/early-releases).
```

File: user/config-imports.md [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+{: data-file=".travis.yml"}
+
+By default, if the same keys are present in multiple imported snippets, they
+are overwritten with the last configuration imported. But you can also specify
+[deep_merge](#merge-mode).
+
+## Importing configs from the same repository
+
+When importing configurations stored in the same repository as your
+`travis.yml`, you can ommit the `<account>/<repository>` part:
```
Comment (Pezi777) [inline user/config-imports.md]:
```suggestion
`travis.yml`, you can commit the `<account>/<repository>` part:
```

File: user/config-imports.md [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+- path/to/other.yml
+```
+{: data-file=".travis.yml"}
+
+The path is relative to the repository's root.
+
+## Importing specific versions of configs
+
+
+Configurations imported from a different repository, fetch the fetch the latest
```
Comment (Pezi777) [inline user/config-imports.md]:
```suggestion
For configurations imported from a different repository, fetch the latest
```

File: user/config-imports.md [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+version of the default branch in the repository.
+
+Configurations imported from the same repository, fetch the commit you are
```
Comment (Pezi777) [inline user/config-imports.md]:
```suggestion
For configurations imported from the same repository, fetch the commit you are
```

File: user/config-imports.md [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+```
+{: data-file=".travis.yml"}
+
+## Merge modes
+
+The merge mode controls how imported configs are being merged (combined) into
+the importing config. A different merge mode can be specified for each imported
+config source.
+
+There are these merge modes:
```
Comment (Pezi777) [inline user/config-imports.md]:
```suggestion
These merge modes are possible:
```

File: user/config-imports.md [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+
+* `merge`
+* `deep_merge`
+* `deep_merge_append` (default)
+* `deep_merge_prepend`
+
+The default merge mode is `deep_merge_append` when using the beta feature
+[build config validation](/user/build-config-validation) that will be rolled
+out to all users over the next couple months, we encourage you to enable it.
```
Comment (Pezi777) [inline user/config-imports.md]:
```suggestion
out to all users over the next couple of months, we encourage you to enable it.
```

File: user/config-imports.md [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+```yaml
+import:
+- source: one.yml
+  mode: merge # shallow merge
+```
+
+### Deep merge
+
+The merge mode `deep_merge` recursively merges sections (keys) that hold maps (hashes),
+but ovewrites sequences (arrays).
```
Comment (Pezi777) [inline user/config-imports.md]:
```suggestion
but overwrites sequences (arrays).
```

File: user/config-imports.md [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+- source: one.yml
+  mode: deep_merge_append
+- source: other.yml
+  mode: deep_merge_prepend
+```
+
+## Private repositories
+
+Before sharing configurations **from** a private repository you need to toggle
+the *Allow importing config files from this repository.* setting in *More
```
Comment (Pezi777) [inline user/config-imports.md]:
```suggestion
the *Allow importing config files from this repository* setting in *More
```

File: user/config-imports.md (hunk 13 above)
Comment (svenfuchs) [reply user/config-imports.md]: I'm not sure this has the same meaning. This has been Sam's (or Maria's?) wording. It is meant to say: 

> If you import a configuration from a different repository, the [version so-and-so] will be fetched/used/imported

File: user/config-imports.md (hunk 14 above)
Comment (svenfuchs) [reply user/config-imports.md]: same as above

File: user/config-imports.md (hunk 15 above)
Comment (svenfuchs) [reply user/config-imports.md]: Hmmmm ... `possible` reads odd to me here. they're not just "possible", but these are all the defined values. i'll think about this more.

File: user/config-imports.md (hunk 13 above)
Comment (svenfuchs) [reply user/config-imports.md]: now made this:

> For configurations imported from a different repository the latest version of
the default branch in the repository will be used by default.

and

> For configurations imported from the same repository the commit you are
currently building will be used by default. This is intended to help while you
are creating and testing the shared configurations.

File: user/build-config-imports.md [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+# or
+
+import:
+  - source: travis-ci/build-configs:rubies.yml@adf1235
+  - source: travis-ci/build-configs:other.yml@v1
+```
+{: data-file=".travis.yml"}
+
+By default, if the same keys are present in multiple imported snippets, they
+are overwritten with the last configuration imported. But you can also specify
```
Comment (Pezi777) [inline user/build-config-imports.md]:
```suggestion
are overwritten with the last imported configuration, but you can also specify
```

File: user/build-config-imports.md [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+```
+{: data-file=".travis.yml"}
+
+## Importing configs from private repositories
+
+In order to share configurations **from** a private repository this needs to
+be allowed on that repository, by enabling the *Allow importing config files from this repository*
+setting in `More options > Settings > Config Import`.
+
+Only private repositories owned by the same organisation or user account will
```
Comment (Pezi777) [inline user/build-config-imports.md]:
```suggestion
> Only private repositories owned by the same organization or user account will
```

File: user/build-config-imports.md [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+be able to import configuration snippets from private repositories. Configs
```
Comment (Pezi777) [inline user/build-config-imports.md]:
```suggestion
> be able to import configuration snippets from private repositories. Configs
```

File: user/build-config-imports.md [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+from private repositories cannot be imported to configs from public
```
Comment (Pezi777) [inline user/build-config-imports.md]:
```suggestion
> from private repositories cannot be imported to configs from public
```

File: user/build-config-imports.md [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+repositories.
```
Comment (Pezi777) [inline user/build-config-imports.md]:
```suggestion
> repositories.
```

**Pull request merged**